### PR TITLE
Add support for SNI when connecting to the instance

### PR DESCRIPTION
**Pull request by diksipav**
I will try to add these tomorrow:

Documentation is missing:
- SNI,
- Cloud Profile,
- tlsCA maybe should be documented too,
- `credentials` is not documented.

**More things to be done**: add SNI to credentials.json in the cli, and later read it from there: `Oh and to address this specifically, although very marginal, we may want to update the CLI to include tls_server_name in credentials.json if --tls-server-name was specified while linking an instance.`

Atm CLOUD_PROFILE is taken from env var when using options. And when using env vars it is not used at all?

### CODE REVIEW

File: packages/driver/src/conUtils.ts [hunk 1]
```diff
@@ -565,6 +587,7 @@ async function parseConnectDsnAndArgs(
       tlsCA: `'tlsCA' option`,
       tlsCAFile: `'tlsCAFile' option`,
       tlsSecurity: `'tlsSecurity' option`,
+      tlsServerName: `tlsServerName option`,
```
Comment (scotttrinh) [inline packages/driver/src/conUtils.ts:590]:
```suggestion
      tlsServerName: `'tlsServerName' option`,
```

File: packages/driver/src/conUtils.ts [hunk 2]
```diff
@@ -619,6 +643,7 @@ async function parseConnectDsnAndArgs(
           secretKey: `'EDGEDB_SECRET_KEY' environment variable`,
           tlsCA: `'EDGEDB_TLS_CA' environment variable`,
           tlsCAFile: `'EDGEDB_TLS_CA_FILE' environment variable`,
+          tlsServerName: `EDGEDB_TLS_SERVER_NAME environment variable`,
```
Comment (scotttrinh) [inline packages/driver/src/conUtils.ts:646]:
```suggestion
          tlsServerName: `'EDGEDB_TLS_SERVER_NAME' environment variable`,
```

File: packages/driver/src/conUtils.ts [hunk 3]
```diff
@@ -663,6 +688,10 @@ async function parseConnectDsnAndArgs(
           .readFileUtf8(stashDir, "database")
           .then((name) => name.trim())
           .catch(() => undefined),
+        serverUtils
+          .readFileUtf8(stashDir, "branch")
+          .then((name) => name.trim())
+          .catch(() => undefined),
```
Comment (scotttrinh) [inline packages/driver/src/conUtils.ts:694]: We're not assigning this to a variable on line 682. Are we missing a test case for this in the shared tests?

Comment (diksipav) [reply packages/driver/src/conUtils.ts:694]: oh yes, u are right

File: packages/driver/src/conUtils.ts [hunk 4]
```diff
@@ -694,13 +694,26 @@ async function parseConnectDsnAndArgs(
           .catch(() => undefined),
       ]);
 
+      let database = _database;
+
+      if (database !== undefined && branch !== undefined) {
+        if (database !== branch) {
+          throw new InterfaceError(
+            "Both database and branch exist in the config dir and don't match.",
```
Comment (scotttrinh) [inline packages/driver/src/conUtils.ts:702]: Do we have a shared client test case for this?

**Review by scotttrinh — APPROVED** [review]

File: packages/driver/src/conUtils.ts (hunk 4 above)
Comment (diksipav) [reply packages/driver/src/conUtils.ts:702]: I don't think that we have for this specific case, but possibillity for this is very rare.

**Review by scotttrinh — APPROVED** [review]

File: packages/driver/src/conUtils.ts (hunk 4 above)
Comment (scotttrinh) [reply packages/driver/src/conUtils.ts:702]: yeah, the test cases cover (currently) rare or impossible cases just to make sure all clients treat different edge cases the same way. probably worth adding a case!
